File: leetcode-bit-manipulation-solutions-enhanced/src/bit_manipulation.cpp
```cpp
#include "bit_manipulation.h"
#include <numeric> // For std::accumulate in SingleNumberIII

// Problem 1: Count Set Bits (Hamming Weight)
namespace CountSetBits {

    /**
     * @brief Counts the number of set bits (1s) in a 32-bit unsigned integer using a simple iterative approach.
     *
     * This method iterates 32 times, checking the least significant bit (LSB) and right-shifting the number.
     *
     * @param n The 32-bit unsigned integer.
     * @return The count of set bits.
     *
     * @complexity
     *   Time: O(k), where k is the number of bits (e.g., 32 for uint32_t).
     *         It performs a fixed number of operations (32 shifts and 32 checks).
     *   Space: O(1)
     */
    int countSetBits_iterative(uint32_t n) {
        int count = 0;
        // Iterate 32 times for a 32-bit integer
        for (int i = 0; i < 32; ++i) {
            // Check if the least significant bit is 1
            if ((n & 1) == 1) {
                count++;
            }
            // Right shift n by 1 to check the next bit
            n >>= 1;
        }
        return count;
    }

    /**
     * @brief Counts the number of set bits (1s) in a 32-bit unsigned integer using Brian Kernighan's Algorithm.
     *
     * This algorithm is more efficient than simple iteration because it only iterates as many times as there are set bits.
     * The operation `n & (n - 1)` clears the least significant set bit in n.
     *
     * @param n The 32-bit unsigned integer.
     * @return The count of set bits.
     *
     * @complexity
     *   Time: O(k), where k is the number of set bits. In the worst case (all bits set), it's O(32).
     *   Space: O(1)
     */
    int countSetBits_brianKernighan(uint32_t n) {
        int count = 0;
        while (n != 0) {
            // Clear the least significant set bit
            n = n & (n - 1);
            count++;
        }
        return count;
    }

    // Lookup table for 8-bit integers (0-255)
    static int lookupTable[256];
    static bool lookupTableInitialized = false;

    /**
     * @brief Initializes the lookup table for counting set bits in bytes.
     *
     * This function should be called once before using `countSetBits_lookupTable`.
     */
    void initializeLookupTable() {
        if (!lookupTableInitialized) {
            for (int i = 0; i < 256; ++i) {
                lookupTable[i] = countSetBits_brianKernighan(i); // Use any method to precompute
            }
            lookupTableInitialized = true;
        }
    }

    /**
     * @brief Counts the number of set bits in a 32-bit unsigned integer using a precomputed lookup table.
     *
     * This method breaks the 32-bit integer into 4 bytes (8 bits each) and sums the precomputed set bit counts
     * for each byte from the lookup table.
     *
     * @param n The 32-bit unsigned integer.
     * @return The count of set bits.
     *
     * @complexity
     *   Time: O(1) after initial lookup table setup. It performs 4 table lookups and 3 additions.
     *   Space: O(256) for the lookup table.
     */
    int countSetBits_lookupTable(uint32_t n) {
        if (!lookupTableInitialized) {
            initializeLookupTable(); // Ensure table is initialized
        }
        int count = 0;
        count += lookupTable[n & 0xFF];          // Count for the first byte (bits 0-7)
        count += lookupTable[(n >> 8) & 0xFF];   // Count for the second byte (bits 8-15)
        count += lookupTable[(n >> 16) & 0xFF];  // Count for the third byte (bits 16-23)
        count += lookupTable[(n >> 24) & 0xFF];  // Count for the fourth byte (bits 24-31)
        return count;
    }

} // namespace CountSetBits


// Problem 2: Single Number III (Find two unique numbers)
namespace SingleNumberIII {

    /**
     * @brief Finds two unique numbers in an array where all other numbers appear twice.
     *
     * This algorithm uses properties of the XOR operator.
     * 1. XOR all numbers together to get `xor_sum = num1 ^ num2`.
     * 2. Find any set bit in `xor_sum`. This bit must be different between `num1` and `num2`.
     *    (e.g., `diff_bit = xor_sum & (-xor_sum)` to get the rightmost set bit).
     * 3. Partition the original array into two groups: those with `diff_bit` set and those without.
     * 4. XOR all numbers in each group. The result of each group's XOR will be one of the unique numbers.
     *
     * @param nums A vector of integers where two numbers appear once and all others twice.
     * @return A vector containing the two unique numbers.
     *
     * @complexity
     *   Time: O(N) because the array is traversed a constant number of times.
     *   Space: O(1) (excluding input/output storage).
     */
    std::vector<int> findTwoUniqueNumbers(std::vector<int>& nums) {
        // Step 1: XOR all numbers to get xor_sum = num1 ^ num2
        int xor_sum = 0;
        for (int num : nums) {
            xor_sum ^= num;
        }

        // Step 2: Find any set bit (distinguishing bit) in xor_sum.
        // We can use `xor_sum & (-xor_sum)` to get the rightmost set bit.
        // This bit is guaranteed to be set in one of the unique numbers and unset in the other.
        int diff_bit = xor_sum & (-xor_sum);

        // Step 3 & 4: Partition numbers into two groups and XOR them
        int num1 = 0; // Will hold one unique number
        int num2 = 0; // Will hold the other unique number

        for (int num : nums) {
            if ((num & diff_bit) != 0) { // If the distinguishing bit is set
                num1 ^= num;
            } else { // If the distinguishing bit is not set
                num2 ^= num;
            }
        }

        return {num1, num2};
    }

} // namespace SingleNumberIII


// Problem 3: Power of Two
namespace PowerOfTwo {

    /**
     * @brief Checks if an integer is a power of two using a loop and division.
     *
     * A number `n` is a power of two if `n > 0` and `n` can be repeatedly divided by 2
     * until it becomes 1, with no remainder at any step.
     *
     * @param n The integer to check.
     * @return True if `n` is a power of two, false otherwise.
     *
     * @complexity
     *   Time: O(log N) as it performs divisions until n becomes 1.
     *   Space: O(1)
     */
    bool isPowerOfTwo_loop(int n) {
        if (n <= 0) {
            return false;
        }
        while (n % 2 == 0) {
            n /= 2;
        }
        return n == 1;
    }

    /**
     * @brief Checks if an integer is a power of two using bit manipulation.
     *
     * A positive integer `n` is a power of two if and only if it has exactly one bit set to 1 in its binary representation.
     * The expression `n & (n - 1)` clears the least significant set bit of `n`.
     * If `n` is a power of two, it has only one set bit. So `n - 1` will have all bits to the right of that set bit flipped
     * from 0 to 1, and that set bit flipped from 1 to 0.
     * Therefore, `n & (n - 1)` will result in 0 if `n` is a power of two.
     *
     * Example:
     *   n = 8 (1000_2)
     *   n-1 = 7 (0111_2)
     *   n & (n-1) = 1000_2 & 0111_2 = 0000_2
     *
     * @param n The integer to check.
     * @return True if `n` is a power of two, false otherwise.
     *
     * @complexity
     *   Time: O(1)
     *   Space: O(1)
     */
    bool isPowerOfTwo_bitManipulation(int n) {
        // A power of two must be positive.
        // For any power of two (e.g., 1, 2, 4, 8, ...), its binary representation
        // has exactly one '1' bit.
        // n & (n - 1) clears the lowest set bit. If n has only one set bit,
        // then n & (n - 1) will be 0.
        return n > 0 && (n & (n - 1)) == 0;
    }

} // namespace PowerOfTwo


// Problem 4: Reverse Bits
namespace ReverseBits {

    /**
     * @brief Reverses the bits of a 32-bit unsigned integer.
     *
     * This method iterates 32 times. In each iteration, it takes the least significant bit (LSB)
     * of the input number `n`, shifts it to the appropriate position in the `reversed_n` result,
     * and adds it (ORs it) to `reversed_n`. Then `n` is right-shifted, and `reversed_n` is left-shifted
     * to prepare for the next bit.
     *
     * @param n The 32-bit unsigned integer to reverse.
     * @return The integer with its bits reversed.
     *
     * @complexity
     *   Time: O(k), where k is the number of bits (32 for uint32_t). Fixed number of operations.
     *   Space: O(1)
     */
    uint32_t reverseBits_iterative(uint32_t n) {
        uint32_t reversed_n = 0;
        for (int i = 0; i < 32; ++i) {
            // Left shift the reversed_n by 1 to make space for the next bit
            reversed_n <<= 1;
            // Get the current LSB of n and add it to reversed_n
            if (n & 1) { // If LSB is 1
                reversed_n |= 1;
            }
            // Right shift n by 1 to process the next bit
            n >>= 1;
        }
        return reversed_n;
    }

} // namespace ReverseBits


// Problem 5: Insert M into N (Update Bits)
namespace UpdateBits {

    /**
     * @brief Inserts integer `m` into integer `n` between bit positions `j` and `i` (inclusive).
     *
     * This problem requires clearing a specific range of bits in `n` and then inserting `m` into that range.
     *
     * Steps:
     * 1. Create a mask to clear bits from `j` down to `i` in `n`.
     *    - Create a sequence of all 1s: `~0` (or `UINT_MAX`).
     *    - Shift `~0` left by `j + 1` to get a mask with 1s from `j+1` to MSB, and 0s from `j` to 0. (e.g., `...11100000`)
     *      `left_mask = ~0 << (j + 1)`
     *    - Create a sequence of 1s from `i-1` down to 0, and 0s elsewhere. (e.g., `000000111...`)
     *      `right_mask = (1 << i) - 1`
     *    - Combine these two to get a mask with 1s outside the range [i, j] and 0s within:
     *      `clear_mask = left_mask | right_mask`
     * 2. Clear the bits in `n` within the range [i, j] using the `clear_mask`.
     *    `n_cleared = n & clear_mask`
     * 3. Shift `m` left by `i` positions to align it with the cleared range.
     *    `m_shifted = m << i`
     * 4. OR `n_cleared` with `m_shifted` to insert `m`.
     *    `result = n_cleared | m_shifted`
     *
     * Example: n = 10000000000_2, m = 10011_2, i = 2, j = 6
     *   n = ...0000 1000 0000 0000_2 (decimal 1024)
     *   m = ...0000 0000 0001 0011_2 (decimal 19)
     *   i = 2, j = 6 (range [2, 6] inclusive)
     *
     *   1. Create clear mask:
     *      left_mask = ~0 << (6 + 1) = ~0 << 7 = ...1111 1000 0000_2
     *      right_mask = (1 << 2) - 1 = 0000 0000 0011_2
     *      clear_mask = left_mask | right_mask = ...1111 1000 0011_2
     *
     *   2. Clear bits in n:
     *      n_cleared = n & clear_mask = 1000 0000 0000_2 & ...1111 1000 0011_2 = 1000 0000 0000_2
     *
     *   3. Shift m to position i:
     *      m_shifted = m << 2 = 0001 0011_2 << 2 = 0100 1100_2
     *
     *   4. Combine:
     *      result = n_cleared | m_shifted = 1000 0000 0000_2 | 0000 0100 1100_2 = 1000 0100 1100_2
     *
     * @param n The target integer.
     * @param m The integer to insert.
     * @param i The starting bit position (inclusive, 0-indexed).
     * @param j The ending bit position (inclusive, 0-indexed).
     * @return The integer `n` with `m` inserted into the specified bit range.
     *
     * @complexity
     *   Time: O(1)
     *   Space: O(1)
     */
    int updateBits(int n, int m, int i, int j) {
        // Handle edge cases for i and j if necessary (e.g., j < i, i or j out of bounds for int)
        // For standard 32-bit int, i and j should be between 0 and 31.
        if (i < 0 || j < i || j >= 32) { // Assuming 32-bit int
            // Or throw an exception, or return an error code
            // For simplicity, we'll proceed assuming valid i,j
        }

        // Step 1: Create a mask to clear bits from j down to i in n.
        // Part 1: Bits from (j+1) to MSB are 1s, others 0s.
        int left_mask = (~0) << (j + 1);

        // Part 2: Bits from 0 to (i-1) are 1s, others 0s.
        // Special case: if i = 0, right_mask should be 0. (1 << 0) - 1 = 0.
        int right_mask = (1 << i) - 1;

        // Combine masks to get a mask with 1s outside [i, j] and 0s inside [i, j].
        int clear_mask = left_mask | right_mask;

        // Step 2: Clear the bits in n within the range [i, j].
        int n_cleared = n & clear_mask;

        // Step 3: Shift m left by i positions to align it with the cleared range.
        int m_shifted = m << i;

        // Step 4: OR n_cleared with m_shifted to insert m.
        return n_cleared | m_shifted;
    }

} // namespace UpdateBits

```

File: coding-interview-bit-manipulation-plus/src/bitManipulation.ts
```typescript
/**
 * src/bitManipulation.ts
 *
 * This file contains implementations of various common bit manipulation problems,
 * each with optimal solutions and, where applicable, alternative approaches.
 * Detailed comments, time/space complexity analysis, and edge case considerations
 * are provided for each function.
 */

/**
 * Problem 1: Counting Set Bits (Hamming Weight)
 *
 * Write a function that takes an unsigned integer and returns the number of '1' bits it has
 * (also known as the Hamming weight).
 *
 * Example:
 * Input: n = 00000000000000000000000000001011 (binary) => 11 (decimal)
 * Output: 3
 */

/**
 * Approach 1.1: Simple Iteration (Bit Shifting)
 *
 * This approach iterates through each bit of the number, checking if the least significant bit (LSB)
 * is 1. It then right-shifts the number to check the next bit.
 *
 * @param n The input unsigned integer.
 * @returns The number of set bits (1s).
 *
 * Time Complexity: O(log N) where N is the value of the number, or O(1) assuming fixed 32-bit integer (32 iterations).
 * Space Complexity: O(1)
 */
export function countSetBitsSimple(n: number): number {
    let count = 0;
    // We assume n is treated as an unsigned 32-bit integer for consistency
    // Although JavaScript numbers are 64-bit floats, bitwise operations
    // treat them as 32-bit signed integers, so we handle the "unsigned"
    // aspect by assuming non-negative input and processing up to 32 bits.
    let num = n >>> 0; // Ensure it's treated as unsigned 32-bit for consistency

    while (num > 0) {
        // Check if the least significant bit is 1
        if ((num & 1) === 1) {
            count++;
        }
        // Right-shift the number by 1 to process the next bit
        num >>>= 1; // Unsigned right shift
    }
    return count;
}

/**
 * Approach 1.2: Brian Kernighan's Algorithm
 *
 * This algorithm is more efficient because it only iterates as many times as there are set bits.
 * It works by repeatedly flipping the least significant set bit to 0. The expression `num & (num - 1)`
 * clears the rightmost set bit in `num`.
 *
 * Example:
 * num = 11 (binary: 0...01011)
 * 1. num = 0...01011, count = 0
 *    num - 1 = 0...01010
 *    num & (num - 1) = 0...01010  (LSB cleared)
 *    count = 1
 * 2. num = 0...01010, count = 1
 *    num - 1 = 0...01001
 *    num & (num - 1) = 0...01000  (LSB cleared)
 *    count = 2
 * 3. num = 0...01000, count = 2
 *    num - 1 = 0...00111
 *    num & (num - 1) = 0...00000  (LSB cleared)
 *    count = 3
 * 4. num = 0, loop terminates.
 *
 * @param n The input unsigned integer.
 * @returns The number of set bits (1s).
 *
 * Time Complexity: O(k) where k is the number of set bits. In the worst case (all bits are 1), it's O(log N) or O(32).
 * Space Complexity: O(1)
 */
export function countSetBitsBrianKernighan(n: number): number {
    let count = 0;
    let num = n >>> 0; // Ensure treated as unsigned 32-bit

    while (num > 0) {
        // This operation clears the rightmost set bit
        num = num & (num - 1);
        count++;
    }
    return count;
}

/**
 * Problem 2: Single Number
 *
 * Given a non-empty array of integers, every element appears twice except for one. Find that single one.
 *
 * Constraints:
 * - Your algorithm should have a linear runtime complexity.
 * - Could you implement it without using extra memory?
 *
 * Example:
 * Input: nums = [2,2,1]
 * Output: 1
 *
 * Input: nums = [4,1,2,1,2]
 * Output: 4
 */

/**
 * Approach 2.1: XOR Property
 *
 * This is the optimal solution leveraging the properties of the XOR bitwise operator:
 * 1. `a ^ 0 = a` (XOR with zero yields the number itself)
 * 2. `a ^ a = 0` (XOR with itself yields zero)
 * 3. `a ^ b ^ a = (a ^ a) ^ b = 0 ^ b = b` (XOR is commutative and associative)
 *
 * By XORing all numbers in the array, all elements that appear twice will cancel each other out
 * (since `a ^ a = 0`), leaving only the single unique number.
 *
 * @param nums An array of integers where one element appears once and others twice.
 * @returns The single unique number.
 *
 * Time Complexity: O(N) where N is the number of elements in the array. We iterate through the array once.
 * Space Complexity: O(1) as no extra memory is used beyond a single variable for the XOR sum.
 */
export function singleNumber(nums: number[]): number {
    let uniqueNum = 0; // Initialize with 0, as x ^ 0 = x

    for (const num of nums) {
        uniqueNum ^= num; // XOR each number with the running XOR sum
    }

    return uniqueNum;
}

/**
 * Approach 2.2: Hash Map (Alternative - Not optimal for constraints)
 *
 * This approach uses a hash map (or Set) to keep track of numbers encountered.
 * If a number is seen for the second time, it's removed from the map.
 * The remaining number in the map is the unique one.
 *
 * This approach does not meet the "without using extra memory" constraint.
 *
 * @param nums An array of integers.
 * @returns The single unique number.
 *
 * Time Complexity: O(N) on average, due to map insertions/deletions. Worst case for hash collisions could be O(N^2) but rare.
 * Space Complexity: O(N) in the worst case (if all numbers are unique until the end), as the map stores elements.
 */
export function singleNumberHashMap(nums: number[]): number {
    const seen = new Set<number>();
    for (const num of nums) {
        if (seen.has(num)) {
            seen.delete(num);
        } else {
            seen.add(num);
        }
    }
    // The only remaining element in the set is the single number.
    // Since constraints state a single number always exists, we can safely
    // assume `seen.values().next().value` will yield the result.
    return seen.values().next().value;
}

/**
 * Problem 3: Power of Two
 *
 * Given an integer `n`, return `true` if it is a power of two. Otherwise, return `false`.
 * An integer `n` is a power of two if there exists an integer `x` such that `n == 2^x`.
 *
 * Example:
 * Input: n = 1 -> Output: true (2^0)
 * Input: n = 16 -> Output: true (2^4)
 * Input: n = 3 -> Output: false
 */

/**
 * Approach 3.1: Bitwise AND Trick
 *
 * A positive integer `n` is a power of two if and only if it has exactly one set bit in its binary representation.
 * For example:
 * 1 (2^0) = 0001
 * 2 (2^1) = 0010
 * 4 (2^2) = 0100
 * 8 (2^3) = 1000
 *
 * If `n` has only one set bit, then `n - 1` will have all bits to the right of that set bit as 1s,
 * and that set bit itself will be 0.
 *
 * Example with `n = 8 (1000b)`:
 * n     = 1000
 * n - 1 = 0111
 * n & (n - 1) = 0000
 *
 * This property `n & (n - 1) === 0` holds true for all powers of two (and only for powers of two)
 * and is a very efficient way to check.
 *
 * Edge Cases:
 * - `n <= 0`: Powers of two are always positive. So, `n` must be greater than 0.
 * - `n = 1`: This is `2^0`, and the condition `1 & (1 - 1) = 1 & 0 = 0` holds.
 *
 * @param n The integer to check.
 * @returns True if n is a power of two, false otherwise.
 *
 * Time Complexity: O(1)
 * Space Complexity: O(1)
 */
export function isPowerOfTwo(n: number): boolean {
    // Powers of two are always positive. 0, negative numbers are not powers of two.
    // And for any power of two 'n', 'n' will have only one bit set.
    // 'n - 1' will have all bits to the right of that set bit as 1s, and that set bit itself as 0.
    // Therefore, 'n & (n - 1)' will be 0.
    return n > 0 && (n & (n - 1)) === 0;
}

/**
 * Approach 3.2: Iterative Division (Alternative)
 *
 * This approach repeatedly divides `n` by 2 until it becomes 1. If at any point `n` is not divisible by 2
 * (i.e., it's odd and not 1), then it's not a power of two.
 *
 * @param n The integer to check.
 * @returns True if n is a power of two, false otherwise.
 *
 * Time Complexity: O(log N) where N is the value of the number.
 * Space Complexity: O(1)
 */
export function isPowerOfTwoIterative(n: number): boolean {
    if (n <= 0) {
        return false; // Powers of two are always positive
    }
    while (n % 2 === 0) {
        n /= 2;
    }
    return n === 1;
}

/**
 * Problem 4: Reverse Bits
 *
 * Reverse bits of a given 32-bit unsigned integer.
 *
 * Example:
 * Input: n = 00000010100101000001111010011100 (binary representation of 43261596)
 * Output: 00111001011110000010100101000000 (binary representation of 964176192)
 *
 * Note:
 * - In C++/Java, the input is typically `uint32_t`. In JavaScript, numbers are 64-bit floats,
 *   but bitwise operators operate on 32-bit signed integers. We'll handle this by assuming
 *   the input is a positive integer that fits within a 32-bit unsigned range and
 *   explicitly using unsigned right shifts.
 */

/**
 * Approach 4.1: Iterative Shifting and Combining
 *
 * This approach iterates 32 times (for a 32-bit integer). In each iteration:
 * 1. It extracts the least significant bit (LSB) of the original number `n`.
 * 2. It shifts the `result` to the left by one position to make space for the new bit.
 * 3. It adds the extracted LSB to the `result`.
 * 4. It right-shifts `n` by one position to process the next bit.
 *
 * Example (for an 8-bit number `n = 00001011`):
 * initial: result = 0, n = 00001011
 *
 * Iteration 1 (i=0):
 *   LSB of n = 1 (00001011 & 1)
 *   result = (0 << 1) | 1 = 00000001
 *   n = 00000101 (n >>> 1)
 *
 * Iteration 2 (i=1):
 *   LSB of n = 1 (00000101 & 1)
 *   result = (00000001 << 1) | 1 = 00000011
 *   n = 00000010
 *
 * Iteration 3 (i=2):
 *   LSB of n = 0 (00000010 & 1)
 *   result = (00000011 << 1) | 0 = 00000110
 *   n = 00000001
 *
 * Iteration 4 (i=3):
 *   LSB of n = 1 (00000001 & 1)
 *   result = (00000110 << 1) | 1 = 00001101
 *   n = 00000000
 *
 * ... continuing for 32 iterations ...
 *
 * After 32 iterations, `result` will contain the reversed bits.
 *
 * @param n The 32-bit unsigned integer to reverse.
 * @returns The integer with its bits reversed.
 *
 * Time Complexity: O(1) (fixed 32 iterations).
 * Space Complexity: O(1)
 */
export function reverseBits(n: number): number {
    let result = 0;
    // We expect a 32-bit unsigned integer. JavaScript's bitwise ops treat numbers as 32-bit signed,
    // so we need to be careful with negative numbers and unsigned shifts.
    // For this problem, 'n' is given as an unsigned integer, so we can assume it's positive.

    for (let i = 0; i < 32; i++) {
        // 1. Shift the current result to the left to make space for the new bit.
        result = result << 1;

        // 2. Extract the least significant bit (LSB) from 'n'.
        //    'n & 1' will be 1 if LSB is 1, and 0 if LSB is 0.
        const lsb = n & 1;

        // 3. Add the extracted LSB to the result.
        //    Using '|' (bitwise OR) effectively sets the LSB of `result` to `lsb`.
        result = result | lsb;

        // 4. Right-shift 'n' by 1 to process the next bit (moving towards LSB).
        //    Using '>>>' (unsigned right shift) ensures that leading bits are filled with zeros,
        //    which is important for maintaining the unsigned nature for large numbers.
        n = n >>> 1;
    }

    // JavaScript bitwise operations produce 32-bit signed integers.
    // If the reversed number has its 31st bit (0-indexed) set, it might be interpreted as negative.
    // To ensure it's treated as unsigned as per the problem, we can use `>>> 0` again,
    // but the problem typically implies the mathematical value of the unsigned number.
    // The `result` variable will correctly hold the value.
    return result >>> 0; // Ensure the final result is treated as unsigned 32-bit.
}

/**
 * Example usage (for local testing or demonstration):
 *
 * Uncomment the block below to run these functions directly.
 */
/*
if (require.main === module) {
    console.log("--- Bit Manipulation Problems ---");

    // Problem 1: Counting Set Bits
    console.log("\nProblem 1: Counting Set Bits");
    const num1 = 11; // Binary: 0...01011
    console.log(`Number: ${num1} (binary: ${num1.toString(2)})`);
    console.log(`  Simple: ${countSetBitsSimple(num1)} set bits`); // Expected: 3
    console.log(`  Kernighan: ${countSetBitsBrianKernighan(num1)} set bits`); // Expected: 3

    const num2 = 255; // Binary: 0...11111111
    console.log(`Number: ${num2} (binary: ${num2.toString(2)})`);
    console.log(`  Simple: ${countSetBitsSimple(num2)} set bits`); // Expected: 8
    console.log(`  Kernighan: ${countSetBitsBrianKernighan(num2)} set bits`); // Expected: 8

    const num3 = 0;
    console.log(`Number: ${num3} (binary: ${num3.toString(2)})`);
    console.log(`  Simple: ${countSetBitsSimple(num3)} set bits`); // Expected: 0
    console.log(`  Kernighan: ${countSetBitsBrianKernighan(num3)} set bits`); // Expected: 0

    // Problem 2: Single Number
    console.log("\nProblem 2: Single Number");
    const arr1 = [2, 2, 1];
    console.log(`Array: [${arr1}] -> Single Number: ${singleNumber(arr1)}`); // Expected: 1
    console.log(`Array: [${arr1}] -> Single Number (HashMap): ${singleNumberHashMap(arr1)}`); // Expected: 1

    const arr2 = [4, 1, 2, 1, 2];
    console.log(`Array: [${arr2}] -> Single Number: ${singleNumber(arr2)}`); // Expected: 4
    console.log(`Array: [${arr2}] -> Single Number (HashMap): ${singleNumberHashMap(arr2)}`); // Expected: 4

    const arr3 = [1];
    console.log(`Array: [${arr3}] -> Single Number: ${singleNumber(arr3)}`); // Expected: 1

    // Problem 3: Power of Two
    console.log("\nProblem 3: Power of Two");
    console.log(`Is 1 a power of two? (Bitwise): ${isPowerOfTwo(1)}`); // Expected: true
    console.log(`Is 1 a power of two? (Iterative): ${isPowerOfTwoIterative(1)}`); // Expected: true
    console.log(`Is 16 a power of two? (Bitwise): ${isPowerOfTwo(16)}`); // Expected: true
    console.log(`Is 16 a power of two? (Iterative): ${isPowerOfTwoIterative(16)}`); // Expected: true
    console.log(`Is 3 a power of two? (Bitwise): ${isPowerOfTwo(3)}`); // Expected: false
    console.log(`Is 3 a power of two? (Iterative): ${isPowerOfTwoIterative(3)}`); // Expected: false
    console.log(`Is 0 a power of two? (Bitwise): ${isPowerOfTwo(0)}`); // Expected: false
    console.log(`Is 0 a power of two? (Iterative): ${isPowerOfTwoIterative(0)}`); // Expected: false
    console.log(`Is -16 a power of two? (Bitwise): ${isPowerOfTwo(-16)}`); // Expected: false

    // Problem 4: Reverse Bits
    console.log("\nProblem 4: Reverse Bits");
    const n1 = 43261596; // 00000010100101000001111010011100
    const reversed1 = reverseBits(n1); // Expected: 964176192 (00111001011110000010100101000000)
    console.log(`Original: ${n1} (binary: ${n1.toString(2).padStart(32, '0')})`);
    console.log(`Reversed: ${reversed1} (binary: ${reversed1.toString(2).padStart(32, '0')})`);

    const n2 = 1; // 0...0001
    const reversed2 = reverseBits(n2); // Expected: 2147483648 (1000...0000)
    console.log(`Original: ${n2} (binary: ${n2.toString(2).padStart(32, '0')})`);
    console.log(`Reversed: ${reversed2} (binary: ${reversed2.toString(2).padStart(32, '0')})`);

    const n3 = 0; // 0...0000
    const reversed3 = reverseBits(n3); // Expected: 0
    console.log(`Original: ${n3} (binary: ${n3.toString(2).padStart(32, '0')})`);
    console.log(`Reversed: ${reversed3} (binary: ${reversed3.toString(2).padStart(32, '0')})`);
}
*/
```